File: C-Plus-Plus/automorphic.cpp
#include <iostream>

/**
 * @brief check if the number is automorphic
 * @param a the natural number to check
 * @return bool true if number is automorphic and its square can fit within
 *          64 bits.
 */
bool automorphic_number(unsigned int a)
{
    unsigned long long b = a * a;

    while (a != 0)
    {
        if (a % 10 != b % 10)
            return false;
        else
        {
            a /= 10;
            b /= 10;
        }
    }

    return true;
}

int main()
{
    unsigned int a;
    std::cin >> a;

    if (automorphic_number(a))
        std::cout << a << " is an automorphic number.\n";
    else
        std::cout << a << " is not an automorphic number.\n";
}

/*
Time complexity: O(log_10 a)
Space complexity: O(2)

An automorphic number (sometimes referred to as a circular number) is a natural number in a given number base b whose square "ends" in the same digits as the number itself.
Here we are considering base 10.

INPUT:
5
OUTPUT:
5 is an automophic number.

INPUT:
25
OUTPUT:
25 is an automophic number.

INPUT:
6
OUTPUT:
6 is not an automophic number.

*/

File: Java/Algorithms/Automorphic.java
package Java.Algorithms;

import java.util.*;

public class Automorphic{
    static boolean isAutomorphic(int n){
        int s=n*n; 
        // Start Comparing digits 
        while(n>0){ 
            // Return false, if any digit of n doesn't match with its square
            if (n%10!=s%10) 
                return false; 
            n/=10; 
            s/=10; 
        } 
        return true; 
    }
    public static void main(String args[]){
        Scanner sc=new Scanner(System.in);
        System.out.print("Enter number: ");
        int n=sc.nextInt();
        if(isAutomorphic(n))
            System.out.println("Automorphic Number");
        else     
            System.out.println("Not an Automorphic Number");
    }
}

/*
    Input:- 
    Enter number: 25
    Output:-
    Automorphic Number
    Logic:-
    25*25=625 which ends with 25
    
    Input:- 
    Enter number: 7
    Output:-
    Not an Automorphic Number
    Logic:-
    7*7=49 which does not end with 7

    Time complexity: O(log n)
    Space complexity: O(1)
*/